plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/organize/imports/ImportsComputer.xtend

```diff
@@ -145,271 +148,213 @@ public class ImportsComputer {
+				resolved.add(key)
 			}
-		];
+		]
+		resolved.forEach [ key |
+			forDisambiguation.removeAll(key)
+		]
 
-		val chosenSolutions = disambiguate(forDisambiguation, interaction);
+		//automatic / semi-automatic / or user only disambiguation
+		val chosenSolutions = DisambiguateUtil.disambiguate(forDisambiguation, interaction,
+			importProvidedElementLabelprovider);
 
-		chosenSolutions.forEach[ret.add(importsFactory.createImport(it, nodelessMarker))];
+		chosenSolutions.forEach[ret.add(importsFactory.createImport(it, contextProject, nodelessMarker))];
 
 		return ret;
 	}
 
-	/** Filters scope by given name and maps result to importable objects. */
-	private def Iterable<ImportableObject> mapToImportableObjects(IScope scopeIdRef, String brokenName) {
-		scopeIdRef
-			.allElements
-			.filter[candidateFilter.apply(it)]
-			.filter[it.name.lastSegment == brokenName]
-			.map[new ImportableObject(brokenName, it, false)]
+	/** Finds unresolved cross references in this script and combines them with provided broken names. 
+	 * @returns list of resolutions for all broken names.
+	 */
+	private def Multimap<String, ImportableObject> createResolutionsForBrokenNames(Script script,
+		IN4JSProject contextProject, Set<String> namesThatWeBroke) {
+		val Multimap<String, ImportableObject> resolutions = LinkedHashMultimap.create();
+
+		val Iterable<ReferenceProxyInfo> unresolved = script.findProxyCrossRefInfo.filter[referenceFilter.test(it)].
+			filter[it.eobject instanceof MemberAccess === false]
+			
+		val brokenNames = new HashSet<String>();
+		brokenNames.addAll(namesThatWeBroke)
+		brokenNames.addAll(unresolved.map[it.name]);
+
+		addResolutionsFromIndex(resolutions, contextProject, brokenNames, script.eResource);
+
+		return resolutions
 	}
 
-	/** Filters scope by the provided name and adds mapped results to the accumulator collection. */
-	private def void addResolutionFromScope(Multimap<String, ImportableObject> resolution, IScope scopeTypeRef, String usedName) {
-		scopeTypeRef
-			.allElements
-			.filter[candidateFilter.apply(it)]
-			.filter[isCandidate(it, usedName)]
-			.forEach[resolution.add(usedName, it)]
+	/**
+	 * Obtains index for based on the provided resource. Matches all broken names against object descriptions in the index.
+	 * Those that pass checks are added to the resolutions map as potential solutions.
+	 */
+	private def void addResolutionsFromIndex(Multimap<String, ImportableObject> resolution,IN4JSProject contextProject,
+		Iterable<String> brokenNames, Resource contextResource) {
+
+		val resourceSet = core.createResourceSet(Optional.fromNullable(contextProject))
+		val resources = core.getXtextIndex(resourceSet).allResourceDescriptions
+		resources.forEach [ res |
+			val candidateProject = core.findProject(res.URI).orNull
+			if (candidateProject !== null) {
+				val isInDeps = ImportSpecifierUtil.getDependencyWithID(candidateProject.projectId, contextProject) !== null
```

Comment:
You can replace the two line with if (ImportSpecifierUtil.getDependencyWithID(candidateProject.projectId, contextProject) !== null)

Reply:
done